src/test/ui/layout/unsafe-cell-hides-niche.rs

```diff
@@ -17,34 +19,51 @@ struct Transparent<T>(T);
+macro_rules! assert_size {
+    ($a:ty, $b:expr) => {{
+        const _: Size::<{$b}> = Size::<{size_of::<$a>()}>;
+    }};
+}
+
+const PTR_SIZE: usize = std::mem::size_of::<*const ()>();
+
 fn main() {
-    assert_eq!(size_of::<Option<Wrapper<u32>>>(),     8);
-    assert_eq!(size_of::<Option<Wrapper<N32>>>(),     4); // (✓ niche opt)
-    assert_eq!(size_of::<Option<Transparent<u32>>>(), 8);
-    assert_eq!(size_of::<Option<Transparent<N32>>>(), 4); // (✓ niche opt)
-    assert_eq!(size_of::<Option<NoNiche<u32>>>(),     8);
-    assert_eq!(size_of::<Option<NoNiche<N32>>>(),     8); // (✗ niche opt)
-
-    assert_eq!(size_of::<Option<UnsafeCell<u32>>>(),  8);
-    assert_eq!(size_of::<Option<UnsafeCell<N32>>>(),  8); // (✗ niche opt)
-
-    assert_eq!(size_of::<       UnsafeCell<&()> >(), 8);
-    assert_eq!(size_of::<Option<UnsafeCell<&()>>>(), 16); // (✗ niche opt)
-    assert_eq!(size_of::<             Cell<&()> >(), 8);
-    assert_eq!(size_of::<Option<      Cell<&()>>>(), 16); // (✗ niche opt)
-    assert_eq!(size_of::<          RefCell<&()> >(), 16);
-    assert_eq!(size_of::<Option<   RefCell<&()>>>(), 24); // (✗ niche opt)
-    assert_eq!(size_of::<           RwLock<&()> >(), 24);
-    assert_eq!(size_of::<Option<    RwLock<&()>>>(), 32); // (✗ niche opt)
-    assert_eq!(size_of::<            Mutex<&()> >(), 16);
-    assert_eq!(size_of::<Option<     Mutex<&()>>>(), 24); // (✗ niche opt)
-
-    assert_eq!(size_of::<       UnsafeCell<&[i32]> >(), 16);
-    assert_eq!(size_of::<Option<UnsafeCell<&[i32]>>>(), 24); // (✗ niche opt)
-    assert_eq!(size_of::<       UnsafeCell<(&(), &())> >(), 16);
-    assert_eq!(size_of::<Option<UnsafeCell<(&(), &())>>>(), 24); // (✗ niche opt)
+    assert_size!(Option<Wrapper<u32>>,     8);
+    assert_size!(Option<Wrapper<N32>>,     4); // (✓ niche opt)
+    assert_size!(Option<Transparent<u32>>, 8);
+    assert_size!(Option<Transparent<N32>>, 4); // (✓ niche opt)
+    assert_size!(Option<NoNiche<u32>>,     8);
+    assert_size!(Option<NoNiche<N32>>,     8); // (✗ niche opt)
+
+    assert_size!(Option<UnsafeCell<u32>>,  8);
+    assert_size!(Option<UnsafeCell<N32>>,  8); // (✗ niche opt)
+
+    assert_size!(       UnsafeCell<&()> , PTR_SIZE);
+    assert_size!(Option<UnsafeCell<&()>>, PTR_SIZE * 2); // (✗ niche opt)
+    assert_size!(             Cell<&()> , PTR_SIZE);
+    assert_size!(Option<      Cell<&()>>, PTR_SIZE * 2); // (✗ niche opt)
+    assert_size!(          RefCell<&()> , PTR_SIZE * 2);
+    assert_size!(Option<   RefCell<&()>>, PTR_SIZE * 3); // (✗ niche opt)
+    assert_size!(           RwLock<&()> , if cfg!(target_pointer_width = "32") { 16 } else { 24 });
+    assert_size!(Option<    RwLock<&()>>, if cfg!(target_pointer_width = "32") { 20 } else { 32 }); // (✗ niche opt)
+    assert_size!(            Mutex<&()> , if cfg!(target_pointer_width = "32") { 12 } else { 16 });
+    assert_size!(Option<     Mutex<&()>>, if cfg!(target_pointer_width = "32") { 16 } else { 24 }); // (✗ niche opt)
```

Comment:
I'm worried this is actually observing weird platform-specific details.
We might want to make a helper function that asserts `Option<T>` is larger than `T` and use it everywhere in this module *before* checking the exact sizes (do they even matter? I guess it's good to have them be tested to some extent).